DEVELOPERS_GUIDE.md

```diff
@@ -0,0 +1,203 @@
+
+## Coding conventions
+
+### Header
+
+Please add the following text to each top each source file:
+
+>
+>/*
+>This file forms part of hpGEM. This package has been developed over a number of years by various people at the University of Twente and a full list of contributors can be found at 
+>http://hpgem.org/about-the-code/team 
+>
+>This code is distributed using BSD 3-Clause License. A copy of which can found below.
+>
+>
+>Copyright (c) 2014, Univesity of Twenete
+>All rights reserved.
+>
+>Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+>
+>1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+>
+>2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+>
+>3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
+>
+>THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A >PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT >LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT >(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+>*/
+
+### Formatting
+
+The code is formatted using [clang-format](https://clang.llvm.org/docs/ClangFormat.html) using the Google formatting conventions. The appropriate ``.clang-format`` is in the main directory. 
+
+
+### General Recommendations
+
+- Any violation to the guide is allowed if it enhances readability.
+
+### Naming Conventions
+
+- File names must be of the form ``FileName``.
+- Directory names must be of the form ``DirectoryName``.
+- Names representing types must be in mixed case starting with upper case, i.e. ``TypeName``.
+- Variable names must be in mixed case starting with lower case, i.e. ``variableName``.
+- Named constants (including enumeration values) must be all uppercase using underscore to separate words.
+- Names representing methods or functions must be verbs and written in mixed case starting with lower case, i.e. ``functionName``.
+- Names representing Classes must be of the form ``ClassName``.
+- Names representing template types should be a single uppercase letter.
```

Comment:
This is quite often violated, with `DIM` as template parameter for the dimension being the canonical example. Maybe we should reformulate this one.